librz/analysis/arch/ppc/ppc_il_ops.c

```diff
@@ -0,0 +1,1587 @@
+	case PPC_INS_MTBR1:
+	case PPC_INS_MTBR2:
+	case PPC_INS_MTBR3:
+	case PPC_INS_MTBR4:
+	case PPC_INS_MTBR5:
+	case PPC_INS_MTBR6:
+	case PPC_INS_MTBR7:
+		NOT_IMPLEMENTED;
+	case PPC_INS_MFXER:
+	case PPC_INS_MTXER:
+		if (id == PPC_INS_MTXER) {
+			return ppc_set_xer(VARG(rS), mode);
+		}
+		NOT_IMPLEMENTED;
+	case PPC_INS_MFDSCR:
+	case PPC_INS_MTDSCR:
+		NOT_IMPLEMENTED;
+	case PPC_INS_MFDSISR:
+	case PPC_INS_MFDAR:
+	case PPC_INS_MFSRR2:
+	case PPC_INS_MFSRR3:
+	case PPC_INS_MFCFAR:
+	case PPC_INS_MFAMR:
+	case PPC_INS_MFPID:
+	case PPC_INS_MFTBLO:
+	case PPC_INS_MFTBHI:
+	case PPC_INS_MFDBATU:
+	case PPC_INS_MFDBATL:
+	case PPC_INS_MFIBATU:
+	case PPC_INS_MFIBATL:
+	case PPC_INS_MFDCCR:
+	case PPC_INS_MFICCR:
+	case PPC_INS_MFDEAR:
+	case PPC_INS_MFESR:
+	case PPC_INS_MFSPEFSCR:
+	case PPC_INS_MFTCR:
+	case PPC_INS_MFASR:
+	case PPC_INS_MFPVR:
+	case PPC_INS_MFTBU:
+	case PPC_INS_MTDSISR:
+	case PPC_INS_MTDAR:
+	case PPC_INS_MTSRR2:
+	case PPC_INS_MTSRR3:
+	case PPC_INS_MTCFAR:
+	case PPC_INS_MTAMR:
+	case PPC_INS_MTPID:
+	case PPC_INS_MTTBL:
+	case PPC_INS_MTTBU:
+	case PPC_INS_MTTBLO:
+	case PPC_INS_MTTBHI:
+	case PPC_INS_MTDBATU:
+	case PPC_INS_MTDBATL:
+	case PPC_INS_MTIBATU:
+	case PPC_INS_MTIBATL:
+	case PPC_INS_MTDCCR:
+	case PPC_INS_MTICCR:
+	case PPC_INS_MTDEAR:
+	case PPC_INS_MTESR:
+	case PPC_INS_MTSPEFSCR:
+	case PPC_INS_MTTCR:
+		NOT_IMPLEMENTED;
+	}
+	if (set_val) {
+		RzILOpEffect *write_spr = ppc_moves_to_spr(id) ? SETG(spr_name, UNSIGNED(size, VARL("val"))) : SETG(rT, EXTZ(VARL("val")));
+		return SEQ2(set_val, write_spr);
+	}
+	return ppc_moves_to_spr(id) ? SETG(spr_name, UNSIGNED(size, VARG(rS))) : SETG(rT, EXTZ(VARG(spr_name)));
+}
+
+/**
+ *
+ * NOTE: Shift instructions are not implemented as in the programmer reference manual.
+ * The manual uses rotate, here we simply use SHIFT ops.
+ */
+static RzILOpEffect *shift_and_rotate(RZ_BORROW csh handle, RZ_BORROW cs_insn *insn, const cs_mode mode) {
+	rz_return_val_if_fail(handle && insn, EMPTY());
+	ut32 id = insn->id;
+	bool sets_cr0 = insn->detail->ppc.update_cr0;
+
+	// READ
+	const char *rA = cs_reg_name(handle, INSOP(0).reg);
+	const char *rS = cs_reg_name(handle, INSOP(1).reg);
+	const char *rB = cs_reg_name(handle, INSOP(2).reg);
+	ut64 sH = INSOP(2).imm;
+	ut64 mB = INSOP(3).imm;
+	ut64 mE = INSOP(4).imm;
+
+	RzILOpPure *n; // Shift/rotate steps
+	RzILOpPure *r; // Rotate result
+	RzILOpPure *b, *e; // Mask begin/end
+	RzILOpPure *into_rA;
+	RzILOpPure *ca_val; // Arithmetic shift instructions set the ca field.
+	RzILOpEffect *set_mask = NULL, *set_ca = NULL, *update_cr0 = NULL;
+
+	// How to read instruction ids:
+	// Letter			Meaning
+	// R/S				Rotate, Shift
+	// L/R				Left, Right
+	// W/D		 		Word, Double Word
+	// A/I				Algebraic, Immediate
+	// C/CL/CR			Clear, clear left/right
+	// M/NM/MI			Mask, AND with mask, mask insert
+
+	switch (id) {
+	default:
+		NOT_IMPLEMENTED;
+	case PPC_INS_ROTLW:
+	case PPC_INS_ROTLWI:
+	case PPC_INS_RLWIMI:
```

Comment:
This can be heavily simplified: `dst = (dst & ~mask) | (rotl32(src, sh) & mask)`
if `mask` is 0, then is just `dst &= ~0`
if `~mask` is 0, then is just `rotl32(src, sh)`
if `sh` is 0, then it can be removed.

Reply:
if you want to implement this, i suggest to test first the mask then the shift.

Reply:
I will look into it. But the first case (mask = 0) is not possible by the ISA. The mask algorithm guaranties to have at least one bit set.